Automattic/wp-calypso
# Plugins: skip SSR rendering if requests take too long

#### Proposed Changes

More info: p58i-d84-p2#comment-55391

This adds a timeout fallback function in case Plugins SSR data requests take too long (more than 2 seconds).
When a timeout happens:
* the page is not SSR-rendered and the client can fill in the data
* the data gets eventually used as it will eventually get cached by react-query
* requests get logged
* bots receive a `504` status code


#### Testing Instructions



* Make sure `plugins/ssr-landing` and `ssr/prefetch-timebox ` flags are enabled
* You'll probably need to [set](https://github.com/Automattic/wp-calypso/pull/68775/files#diff-b4e8533bd35c0ed97c4d6960f21e96e6987d8c8e1cb611130945ca757ed0508aR11) the `PREFETCH_TIMEOUT` to a very low value to trigger the timeout scenario
* Run your instance with a `CALYPSO_LOGFILE=errors.log` env variable

* Log out and disable JavaScript
* Go to `/plugins` 
* You should see the empty page
<img width="320" alt="Screenshot on 2022-10-07 at 18-08-32" src="https://user-images.githubusercontent.com/2749938/194586623-df6fae73-3585-45fc-aabb-66c5a924c738.png">

* Refresh
* On the second request you should be able to see the Plugins landing page with the 3 sections Paid, Editor's Pick, and Free
<img width="320" alt="Screenshot on 2022-10-07 at 18-08-44" src="https://user-images.githubusercontent.com/2749938/194586684-75a362fd-5f27-4258-b562-62f0373ca7be.png">

* Searching inside the `errors.log` file, you should find an entry with a `plugins prefetch timeout` message
* Spoofing your User Agent as a crawler bot should result in a `504` response
<img width="320" alt="Screenshot on 2022-10-07 at 18-08-23" src="https://user-images.githubusercontent.com/2749938/194586708-84a0924b-9fa5-494f-886a-0b8a5df018bc.png">
